File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/components/MediaTypeGlyph.kt
package com.rork.guidestreamtvandroid.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.foundation.layout.size
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.RoundRect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * The movie / series glyph beside a title in the home rails (GUI-70).
 *
 * Hand-drawn rather than a Material icon on purpose: Material's `Tv` is a flat
 * panel and `LiveTv` draws signal waves, while SF Symbols has no antenna
 * television at all — three platforms could not have matched any other way.
 *
 * The twin lives in ios/Shared/MediaTypeGlyph.swift and traces the same
 * 24-unit grid, coordinate for coordinate. Change one, change the other.
 */
@Composable
fun MediaTypeGlyph(
    isTV: Boolean,
    modifier: Modifier = Modifier,
    size: Dp = 12.dp,
    color: Color = Color.White.copy(alpha = 0.62f),
) {
    Canvas(modifier = modifier.size(size)) {
        val s = kotlin.math.min(this.size.width, this.size.height) / 24f
        val stroke = Stroke(
            // 1.7 at 24 units, scaled so the shape keeps its proportions
            // instead of going spindly at 12dp.
            width = 1.7f * s,
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
        )
        val path = Path()

        if (isTV) {
            // Television with rabbit-ear antennae.
            path.addRoundRect(
                RoundRect(
                    left = 2.5f * s, top = 9f * s, right = 21.5f * s, bottom = 21f * s,
                    cornerRadius = CornerRadius(2.5f * s, 2.5f * s),
                )
            )
            path.moveTo(8f * s, 9f * s); path.lineTo(5f * s, 3.5f * s)
            path.moveTo(16f * s, 9f * s); path.lineTo(19f * s, 3.5f * s)
        } else {
            // Film strip: a body with a sprocket column down each side.
            path.addRoundRect(
                RoundRect(
                    left = 2.5f * s, top = 4.5f * s, right = 21.5f * s, bottom = 19.5f * s,
                    cornerRadius = CornerRadius(2.5f * s, 2.5f * s),
                )
            )
            for (x in listOf(7f, 17f)) {
                path.moveTo(x * s, 4.5f * s); path.lineTo(x * s, 19.5f * s)
            }
            for (y in listOf(9.5f, 14.5f)) {
                path.moveTo(2.5f * s, y * s); path.lineTo(7f * s, y * s)
                path.moveTo(17f * s, y * s); path.lineTo(21.5f * s, y * s)
            }
        }

        drawPath(path = path, color = color, style = stroke)
    }
}

File: ios/Shared/MediaTypeGlyph.swift
//
//  MediaTypeGlyph.swift
//  GuideStreamTV / GuideStreamTVTV
//
//  The movie / series glyph that sits beside a title in the home rails
//  (GUI-70). A member of the Shared group, so iOS and tvOS draw the identical
//  shape from one definition.
//
//  Hand-drawn rather than SF Symbols on purpose: SF Symbols has no television
//  with an antenna — `tv` is a flat panel on a stand — and Android's Material
//  set draws both shapes differently again. Three platforms could not have
//  looked the same any other way. The Compose twin lives in
//  android/.../ui/components/MediaTypeGlyph.kt and traces the same 24-unit
//  grid; change one, change the other.
//

import SwiftUI

/// Which glyph to draw. Deliberately not `Bool` at the call site — `isTV` reads
/// as a data flag, this reads as a picture.
enum MediaTypeGlyphKind {
    case movie
    case series

    init(isTV: Bool) {
        self = isTV ? .series : .movie
    }
}

/// A film strip (movie) or an antenna television (series), stroked on a
/// 24×24 grid and scaled to `size`. Sized in points so it tracks Dynamic Type
/// when the caller scales it.
struct MediaTypeGlyph: View {
    let kind: MediaTypeGlyphKind
    var size: CGFloat = 12
    var color: Color = Color.white.opacity(0.62)

    /// Stroke weight at 24pt, scaled with the glyph so the shape keeps its
    /// proportions instead of going spindly at small sizes.
    private var lineWidth: CGFloat { 1.7 * (size / 24) }

    var body: some View {
        Canvas { context, canvasSize in
            let s = min(canvasSize.width, canvasSize.height) / 24
            let stroke = StrokeStyle(
                lineWidth: lineWidth,
                lineCap: .round,
                lineJoin: .round
            )
            context.stroke(
                path(scale: s),
                with: .color(color),
                style: stroke
            )
        }
        .frame(width: size, height: size)
        // The glyph repeats the media type already announced by the title's
        // own accessibility context; announcing it twice is noise.
        .accessibilityHidden(true)
    }

    private func path(scale s: CGFloat) -> Path {
        var p = Path()
        switch kind {
        case .movie:
            // Film strip: a body with a sprocket column down each side.
            p.addRoundedRect(
                in: CGRect(x: 2.5 * s, y: 4.5 * s, width: 19 * s, height: 15 * s),
                cornerSize: CGSize(width: 2.5 * s, height: 2.5 * s)
            )
            for x in [7.0, 17.0] {
                p.move(to: CGPoint(x: x * s, y: 4.5 * s))
                p.addLine(to: CGPoint(x: x * s, y: 19.5 * s))
            }
            for y in [9.5, 14.5] {
                p.move(to: CGPoint(x: 2.5 * s, y: y * s))
                p.addLine(to: CGPoint(x: 7 * s, y: y * s))
                p.move(to: CGPoint(x: 17 * s, y: y * s))
                p.addLine(to: CGPoint(x: 21.5 * s, y: y * s))
            }
        case .series:
            // Television with rabbit-ear antennae.
            p.addRoundedRect(
                in: CGRect(x: 2.5 * s, y: 9 * s, width: 19 * s, height: 12 * s),
                cornerSize: CGSize(width: 2.5 * s, height: 2.5 * s)
            )
            p.move(to: CGPoint(x: 8 * s, y: 9 * s))
            p.addLine(to: CGPoint(x: 5 * s, y: 3.5 * s))
            p.move(to: CGPoint(x: 16 * s, y: 9 * s))
            p.addLine(to: CGPoint(x: 19 * s, y: 3.5 * s))
        }
        return p
    }
}

extension MediaTypeGlyph {
    init(isTV: Bool, size: CGFloat = 12, color: Color = Color.white.opacity(0.62)) {
        self.init(kind: MediaTypeGlyphKind(isTV: isTV), size: size, color: color)
    }
}
